src/PluginClient.php

```diff
@@ -79,7 +78,7 @@ public function sendRequest(RequestInterface $request)
         $pluginChain = $this->createPluginChain($this->plugins, function (RequestInterface $request) {
             try {
                 return new FulfilledPromise($this->client->sendRequest($request));
-            } catch (HttplugException $exception) {
+            } catch (\Exception $exception) {
```

Comment:
> Make a BC break and allows everything as a rejected or resolved value and add checks everywhere where we use promise

But allowing everything from the promise point of view is what we have at the moment. However, during the HTTP request flow, the promise should only receive HTTP exceptions IMO. Non-HTTP exceptions (either emitted by plugins or not) should not be caught in the first place.

The other thing is that domain related exceptions should probably not be thrown during the HTTP flow, but after it is finished in the API Client class or whatever consumes the HTTP endpoint. Even if they should be done for all requests (eg. rate limiting).

May this be the root of the problem? https://github.com/php-http/guzzle6-adapter/blob/master/src/Promise.php#L65

The phpdoc says the stored exception should be an HttplugException, but we store runtime ones there too?

About HttpPromise: 

1. It is BC break (EVERY interface change is BC break)
2. Extending the Promise interface doesn't help as there is no return type overloading (even if it is only typehint, but think about it as PHP 7 return typehints are implemented)

What I would do: document better and find why a non-http exception is not thrown immediately.